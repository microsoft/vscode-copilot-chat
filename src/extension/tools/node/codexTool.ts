/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import type * as vscode from 'vscode';
import { LanguageModelToolResult, MarkdownString } from '../../../vscodeTypes';
import { ToolName } from '../common/toolNames';
import { ICopilotTool, ToolRegistry } from '../common/toolsRegistry';
import { ReadFileParams } from './readFileTool';

export class CodexTool implements ICopilotTool<ReadFileParams> {
	public static toolName = ToolName.ConfirmationTool;

	constructor() { }

	async invoke(options: vscode.LanguageModelToolInvocationOptions<any>, token: vscode.CancellationToken) {
		return new LanguageModelToolResult([]);
	}

	prepareInvocation(options: vscode.LanguageModelToolInvocationPrepareOptions<any>, token: vscode.CancellationToken): vscode.ProviderResult<vscode.PreparedToolInvocation> {
		// return new PreparedTerminalToolInvocation(options.input.detail,
		// 	'sh',
		// 	{
		// 		title: options.input.message,
		// 		message: ''
		// 	}
		// );
		return {
			confirmationMessages: {
				title: options.input.message,
				message: new MarkdownString(options.input.detail),
			}
		};
	}
}

ToolRegistry.registerTool(CodexTool);
