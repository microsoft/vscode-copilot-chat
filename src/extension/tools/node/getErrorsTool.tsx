/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import * as l10n from '@vscode/l10n';
import { BasePromptElementProps, PromptElement, PromptElementProps } from '@vscode/prompt-tsx';
import type * as vscode from 'vscode';
import { ILanguageDiagnosticsService } from '../../../platform/languages/common/languageDiagnosticsService';
import { ILogService } from '../../../platform/log/common/logService';
import { INotebookService } from '../../../platform/notebook/common/notebookService';
import { IPromptPathRepresentationService } from '../../../platform/prompts/common/promptPathRepresentationService';
import { IWorkspaceService } from '../../../platform/workspace/common/workspaceService';
import { getLanguage } from '../../../util/common/languages';
import { findNotebook } from '../../../util/common/notebooks';
import { isLocation } from '../../../util/common/types';
import { coalesce } from '../../../util/vs/base/common/arrays';
import { CancellationToken } from '../../../util/vs/base/common/cancellation';
import { Disposable } from '../../../util/vs/base/common/lifecycle';
import { isEqualOrParent } from '../../../util/vs/base/common/resources';
import { URI } from '../../../util/vs/base/common/uri';
import { IInstantiationService } from '../../../util/vs/platform/instantiation/common/instantiation';
import { DiagnosticSeverity, ExtendedLanguageModelToolResult, LanguageModelPromptTsxPart, MarkdownString, Range } from '../../../vscodeTypes';
import { IBuildPromptContext } from '../../prompt/common/intents';
import { renderPromptElementJSON } from '../../prompts/node/base/promptRenderer';
import { Tag } from '../../prompts/node/base/tag';
import { DiagnosticContext, Diagnostics } from '../../prompts/node/inline/diagnosticsContext';
import { ToolName } from '../common/toolNames';
import { ICopilotTool, ToolRegistry } from '../common/toolsRegistry';
import { checkCancellation, formatUriForFileWidget, resolveToolInputPath } from './toolUtils';

interface IGetErrorsParams {
	// Note that empty array is not the same as absence; empty array
	// will not return any errors. Absence returns all errors.
	filePaths?: string[];
	// sparse array of ranges, as numbers because it goes through JSON
	// ignored if filePaths is missing / null.
	ranges?: ([a: number, b: number, c: number, d: number] | undefined)[];
}

export class GetErrorsTool extends Disposable implements ICopilotTool<IGetErrorsParams> {
	public static readonly toolName = ToolName.GetErrors;

	constructor(
		@IInstantiationService private readonly instantiationService: IInstantiationService,
		@ILanguageDiagnosticsService private readonly languageDiagnosticsService: ILanguageDiagnosticsService,
		@IWorkspaceService private readonly workspaceService: IWorkspaceService,
		@IPromptPathRepresentationService private readonly promptPathRepresentationService: IPromptPathRepresentationService,
		@INotebookService private readonly notebookService: INotebookService,
		@ILogService private readonly logService: ILogService
	) {
		super();
	}

	/**
	 * Get diagnostics for the given paths and optional ranges.
	 * Note - This is made public for testing purposes only.
	 */
	public getDiagnostics(paths: { uri: URI; range: Range | undefined }[]): Array<{ uri: URI; diagnostics: vscode.Diagnostic[] }> {
		const results: Array<{ uri: URI; diagnostics: vscode.Diagnostic[] }> = [];

		// for notebooks, we need to find the cell matching the range and get diagnostics for that cell
		const nonNotebookPaths = paths.filter(p => {
			const isNotebook = this.notebookService.hasSupportedNotebooks(p.uri);
			if (isNotebook) {
				const diagnostics = this.getNotebookCellDiagnostics(p.uri);
				results.push({ uri: p.uri, diagnostics });
			}

			return !isNotebook;
		});

		if (nonNotebookPaths.length === 0) {
			return results;
		}

		const pendingMatchPaths = new Set(nonNotebookPaths.map(p => p.uri));

		// for non-notebooks, we get all diagnostics and filter down
		for (const [resource, entries] of this.languageDiagnosticsService.getAllDiagnostics()) {
			const pendingDiagnostics = entries.filter(d => d.severity <= DiagnosticSeverity.Warning);

			// find all path&range pairs and collect the ranges to further filter diagnostics
			// if any path matches the resource without a range, take all diagnostics for that file
			// otherwise, filter diagnostics to those intersecting one of the provided ranges
			const ranges: Range[] = [];
			let shouldTakeAll = false;
			let foundMatch = false;
			for (const path of nonNotebookPaths) {
				// we support file or folder paths
				if (isEqualOrParent(resource, path.uri)) {
					foundMatch = true;

					if (pendingMatchPaths.has(path.uri)) {
						pendingMatchPaths.delete(path.uri);
					}

					if (path.range) {
						ranges.push(path.range);
					} else {
						// no range, so all diagnostics for this file
						shouldTakeAll = true;
						break;
					}
				}
			}

			if (shouldTakeAll) {
				results.push({ uri: resource, diagnostics: pendingDiagnostics });
				continue;
			}

			if (foundMatch && ranges.length > 0) {
				const diagnostics = pendingDiagnostics.filter(d => ranges.some(range => d.range.intersection(range)));
				results.push({ uri: resource, diagnostics });
			}
		}

		// for any given paths that didn't match any files, return empty diagnostics for each of them
		for (const uri of pendingMatchPaths) {
			results.push({ uri, diagnostics: [] });
		}

		return results;
	}

	async invoke(options: vscode.LanguageModelToolInvocationOptions<IGetErrorsParams>, token: CancellationToken) {
		const getAll = () => this.languageDiagnosticsService.getAllDiagnostics()
			.map(d => ({ uri: d[0], diagnostics: d[1].filter(e => e.severity <= DiagnosticSeverity.Warning) }))
			// filter any documents w/o warnings or errors
			.filter(d => d.diagnostics.length > 0);

		const getSome = (filePaths: string[]) =>
			this.getDiagnostics(filePaths.map((filePath, i) => {
				const uri = resolveToolInputPath(filePath, this.promptPathRepresentationService);
				const range = options.input.ranges?.[i];
				if (!uri) {
					throw new Error(`Invalid input path ${filePath}`);
				}

				return { uri, range: range ? new Range(...range) : undefined };
			}));

		const ds = options.input.filePaths?.length ? getSome(options.input.filePaths) : getAll();

		const diagnostics = coalesce(await Promise.all(ds.map((async ({ uri, diagnostics }) => {
			try {
				const document = await this.workspaceService.openTextDocumentAndSnapshot(uri);
				checkCancellation(token);
				return {
					uri,
					diagnostics,
					context: { document, language: getLanguage(document) }
				};
			} catch (e) {
				this.logService.error(e, 'get_errors failed to open doc with diagnostics');
				return undefined;
			}
		}))));
		checkCancellation(token);

		const result = new ExtendedLanguageModelToolResult([
			new LanguageModelPromptTsxPart(
				await renderPromptElementJSON(this.instantiationService, DiagnosticToolOutput, { diagnosticsGroups: diagnostics, maxDiagnostics: 50 }, options.tokenizationOptions, token)
			)
		]);

		const numDiagnostics = diagnostics.reduce((acc, { diagnostics }) => acc + diagnostics.length, 0);
		const formattedURIs = this.formatURIs(diagnostics.map(d => d.uri));
		if (options.input.filePaths?.length) {
			result.toolResultMessage = numDiagnostics === 0 ?
				new MarkdownString(l10n.t`Checked ${formattedURIs}, no problems found`) :
				numDiagnostics === 1 ?
					new MarkdownString(l10n.t`Checked ${formattedURIs}, 1 problem found`) :
					new MarkdownString(l10n.t`Checked ${formattedURIs}, ${numDiagnostics} problems found`);
		} else {
			result.toolResultMessage = numDiagnostics === 0 ?
				new MarkdownString(l10n.t`Checked workspace, no problems found`) :
				numDiagnostics === 1 ?
					new MarkdownString(l10n.t`Checked workspace, 1 problem found in ${formattedURIs}`) :
					new MarkdownString(l10n.t`Checked workspace, ${numDiagnostics} problems found in ${formattedURIs}`);
		}

		return result;
	}

	prepareInvocation(options: vscode.LanguageModelToolInvocationPrepareOptions<IGetErrorsParams>, token: vscode.CancellationToken): vscode.ProviderResult<vscode.PreparedToolInvocation> {
		if (!options.input.filePaths?.length) {
			// When no file paths provided, check all files with diagnostics
			return {
				invocationMessage: new MarkdownString(l10n.t`Checking workspace for problems`),
			};
		}
		else {
			const uris = options.input.filePaths.map(filePath => resolveToolInputPath(filePath, this.promptPathRepresentationService));
			if (uris.some(uri => uri === undefined)) {
				throw new Error('Invalid file path provided');
			}

			return {
				invocationMessage: new MarkdownString(l10n.t`Checking ${this.formatURIs(uris)}`),
			};
		}
	}

	private formatURIs(uris: URI[]): string {
		return uris.map(formatUriForFileWidget).join(', ');
	}

	private getNotebookCellDiagnostics(uri: URI) {
		const notebook = findNotebook(uri, this.workspaceService.notebookDocuments);

		if (!notebook) {
			this.logService.error(`Notebook not found: ${uri.toString()}, could not retrieve diagnostics`);
			return [];
		}

		return notebook.getCells()
			.flatMap((cell) => {
				const uri = cell.document.uri;
				return this.languageDiagnosticsService.getDiagnostics(uri);
			});
	}

	async provideInput(promptContext: IBuildPromptContext): Promise<IGetErrorsParams | undefined> {
		const seen = new Set<string>();

		const filePaths: string[] = [];
		const ranges: ([a: number, b: number, c: number, d: number] | undefined)[] = [];

		function addPath(path: string, range: vscode.Range | undefined) {
			if (!seen.has(path)) {
				seen.add(path);
				filePaths.push(path);
				ranges.push(range && [range.start.line, range.start.character, range.end.line, range.end.character]);
			}
		}

		for (const ref of promptContext.chatVariables) {
			if (URI.isUri(ref.value)) {
				addPath(this.promptPathRepresentationService.getFilePath(ref.value), undefined);
			} else if (isLocation(ref.value)) {
				addPath(this.promptPathRepresentationService.getFilePath(ref.value.uri), ref.value.range);
			}
		}

		if (promptContext.workingSet) {
			for (const file of promptContext.workingSet) {
				addPath(this.promptPathRepresentationService.getFilePath(file.document.uri), file.range);
			}
		}

		if (!filePaths.length) {
			for (const [uri, diags] of this.languageDiagnosticsService.getAllDiagnostics()) {
				const path = this.promptPathRepresentationService.getFilePath(uri);
				if (diags.length) {
					let range = diags[0].range;
					for (let i = 1; i < diags.length; i++) {
						range = range.union(diags[i].range);
					}
					addPath(path, range);
				}
			}
		}

		return {
			filePaths,
			ranges
		};
	}
}

ToolRegistry.registerTool(GetErrorsTool);

interface IDiagnosticToolOutputProps extends BasePromptElementProps {
	diagnosticsGroups: { context: DiagnosticContext; uri: URI; diagnostics: vscode.Diagnostic[] }[];
	maxDiagnostics?: number;
}

export class DiagnosticToolOutput extends PromptElement<IDiagnosticToolOutputProps> {
	constructor(
		props: PromptElementProps<IDiagnosticToolOutputProps>,
		@IPromptPathRepresentationService private readonly promptPathRepresentationService: IPromptPathRepresentationService,
	) {
		super(props);
	}

	render() {
		if (!this.props.diagnosticsGroups.length) {
			return <>No errors found.</>;
		}

		let diagnosticsGroups = this.props.diagnosticsGroups;
		let limitMsg;
		if (typeof this.props.maxDiagnostics === 'number') {
			let remaining = this.props.maxDiagnostics;
			diagnosticsGroups = this.props.diagnosticsGroups.map(group => {
				if (remaining <= 0) {
					return { ...group, diagnostics: [] };
				}
				const take = Math.min(group.diagnostics.length, remaining);
				remaining -= take;
				return { ...group, diagnostics: group.diagnostics.slice(0, take) };
			});
			const totalDiagnostics = this.props.diagnosticsGroups.reduce((acc, group) => acc + group.diagnostics.length, 0);
			limitMsg = totalDiagnostics > this.props.maxDiagnostics
				? <>Showing first {this.props.maxDiagnostics} results out of {totalDiagnostics}<br /></>
				: undefined;
		}

		return <>
			{limitMsg}
			{diagnosticsGroups.map(d =>
				<Tag name='errors' attrs={{ path: this.promptPathRepresentationService.getFilePath(d.uri) }}>
					{d.diagnostics.length
						? <Diagnostics
							documentContext={d.context}
							diagnostics={d.diagnostics}
							includeRelatedInfos={false} // avoid blowing up the prompt #12655
						/>
						: 'No errors found'}
				</Tag>
			)}
		</>;
	}
}
