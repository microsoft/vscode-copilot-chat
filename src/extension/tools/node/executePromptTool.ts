/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import type * as vscode from 'vscode';
import { IInstantiationService } from '../../../util/vs/platform/instantiation/common/instantiation';
import { ChatResponseMarkdownPart, ExtendedLanguageModelToolResult, LanguageModelTextPart } from '../../../vscodeTypes';
import { Conversation, Turn } from '../../prompt/common/conversation';
import { IBuildPromptContext } from '../../prompt/common/intents';
import { ExecutePromptToolCallingLoop } from '../../prompt/node/executePromptToolCalling';
import { ToolName } from '../common/toolNames';
import { CopilotToolMode, ICopilotTool, ToolRegistry } from '../common/toolsRegistry';
import { ChatResponseStreamImpl } from '../../../util/common/chatResponseStreamImpl';

export interface IExecuteTaskParams {
	prompt: string;
	description: string;
}

class ExecuteTaskTool implements ICopilotTool<IExecuteTaskParams> {
	public static readonly toolName = ToolName.ExecuteTask;
	private _inputContext: IBuildPromptContext | undefined;

	constructor(
		@IInstantiationService private readonly instantiationService: IInstantiationService,
	) { }

	async invoke(options: vscode.LanguageModelToolInvocationOptions<IExecuteTaskParams>, token: vscode.CancellationToken) {

		const loop = this.instantiationService.createInstance(ExecutePromptToolCallingLoop, {
			toolCallLimit: 25,
			conversation: new Conversation('', [new Turn('', { type: 'user', message: options.input.prompt })]),
			request: this._inputContext!.request!,
			location: this._inputContext!.request!.location,
			promptText: options.input.prompt,
		});

		// TODO This also prevents codeblock pills from being rendered
		// I want to render this content as thinking blocks but couldn't get it to work
		const stream = this._inputContext?.stream && ChatResponseStreamImpl.filter(
			this._inputContext.stream,
			part => !(part instanceof ChatResponseMarkdownPart)
		);

		const loopResult = await loop.run(stream, token);
		// Return the text of the last assistant response from the tool calling loop
		const lastRoundResponse = loopResult.toolCallRounds.at(-1)?.response ?? loopResult.round.response ?? '';
		const result = new ExtendedLanguageModelToolResult([new LanguageModelTextPart(lastRoundResponse)]);
		return result;
	}

	prepareInvocation(options: vscode.LanguageModelToolInvocationPrepareOptions<IExecuteTaskParams>, token: vscode.CancellationToken): vscode.ProviderResult<vscode.PreparedToolInvocation> {
		const { input } = options;
		try {
			return {
				invocationMessage: input.description,
			};
		} catch {
			return;
		}
	}

	async resolveInput(input: IExecuteTaskParams, promptContext: IBuildPromptContext, mode: CopilotToolMode): Promise<IExecuteTaskParams> {
		this._inputContext = promptContext;
		return input;
	}
}

ToolRegistry.registerTool(ExecuteTaskTool);
