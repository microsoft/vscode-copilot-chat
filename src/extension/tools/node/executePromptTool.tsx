/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import * as l10n from '@vscode/l10n';
import type * as vscode from 'vscode';
import { IPromptPathRepresentationService } from '../../../platform/prompts/common/promptPathRepresentationService';
import { IWorkspaceService } from '../../../platform/workspace/common/workspaceService';
import { IInstantiationService } from '../../../util/vs/platform/instantiation/common/instantiation';
import { ExtendedLanguageModelToolResult, LanguageModelTextPart, MarkdownString } from '../../../vscodeTypes';
import { IBuildPromptContext } from '../../prompt/common/intents';
import { ExecutePromptToolCallingLoop } from '../../prompt/node/executePromptToolCalling';
import { ToolName } from '../common/toolNames';
import { CopilotToolMode, ICopilotTool, ToolRegistry } from '../common/toolsRegistry';
import { assertFileOkForTool, formatUriForFileWidget, resolveToolInputPath } from './toolUtils';

export interface IExecutePromptParams {
	filePath: string;
}

class ExecutePromptTool implements ICopilotTool<IExecutePromptParams> {
	public static readonly toolName = ToolName.ExecutePrompt;
	private _inputContext: IBuildPromptContext | undefined;

	constructor(
		@IInstantiationService private readonly instantiationService: IInstantiationService,
		@IWorkspaceService private readonly workspaceService: IWorkspaceService,
		@IPromptPathRepresentationService private readonly promptPathRepresentationService: IPromptPathRepresentationService,
	) { }

	async invoke(options: vscode.LanguageModelToolInvocationOptions<IExecutePromptParams>, token: vscode.CancellationToken) {
		if (!options.input.filePath) {
			throw new Error('Invalid input');
		}

		// Read the prompt file as text and include a reference
		const uri = resolveToolInputPath(options.input.filePath, this.promptPathRepresentationService);
		await this.instantiationService.invokeFunction(accessor => assertFileOkForTool(accessor, uri));
		const doc = await this.workspaceService.openTextDocument(uri);
		const promptText = doc.getText();

		const loop = this.instantiationService.createInstance(ExecutePromptToolCallingLoop, {
			toolCallLimit: 5,
			conversation: this._inputContext!.conversation!,
			request: this._inputContext!.request!,
			location: this._inputContext!.request!.location,
			promptText,
		});

		const loopResult = await loop.run(this._inputContext?.stream, token);
		// Return the text of the last assistant response from the tool calling loop
		const lastRoundResponse = loopResult.toolCallRounds.at(-1)?.response ?? loopResult.round.response ?? '';
		const result = new ExtendedLanguageModelToolResult([new LanguageModelTextPart(lastRoundResponse)]);
		return result;
	}

	prepareInvocation(options: vscode.LanguageModelToolInvocationPrepareOptions<IExecutePromptParams>, token: vscode.CancellationToken): vscode.ProviderResult<vscode.PreparedToolInvocation> {
		const { input } = options;
		if (!input.filePath) {
			return;
		}
		try {
			const uri = resolveToolInputPath(input.filePath, this.promptPathRepresentationService);
			return {
				invocationMessage: new MarkdownString(l10n.t`Executing prompt file ${formatUriForFileWidget(uri)}`),
				pastTenseMessage: new MarkdownString(l10n.t`Executed prompt file ${formatUriForFileWidget(uri)}`),
			};
		} catch {
			return;
		}
	}

	async resolveInput(input: IExecutePromptParams, promptContext: IBuildPromptContext, mode: CopilotToolMode): Promise<IExecutePromptParams> {
		this._inputContext = promptContext;
		return input;
	}
}

ToolRegistry.registerTool(ExecutePromptTool);
