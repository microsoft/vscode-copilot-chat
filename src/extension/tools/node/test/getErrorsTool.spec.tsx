/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import { afterEach, beforeEach, expect, suite, test } from 'vitest';
import { ILanguageDiagnosticsService } from '../../../../platform/languages/common/languageDiagnosticsService';
import { TestLanguageDiagnosticsService } from '../../../../platform/languages/common/testLanguageDiagnosticsService';
import { IPromptPathRepresentationService } from '../../../../platform/prompts/common/promptPathRepresentationService';
import { ITestingServicesAccessor, TestingServiceCollection } from '../../../../platform/test/node/services';
import { TestWorkspaceService } from '../../../../platform/test/node/testWorkspaceService';
import { IWorkspaceService } from '../../../../platform/workspace/common/workspaceService';
import { createTextDocumentData } from '../../../../util/common/test/shims/textDocument';
import { CancellationToken } from '../../../../util/vs/base/common/cancellation';
import { URI } from '../../../../util/vs/base/common/uri';
import { SyncDescriptor } from '../../../../util/vs/platform/instantiation/common/descriptors';
import { IInstantiationService } from '../../../../util/vs/platform/instantiation/common/instantiation';
import { DiagnosticSeverity, Range } from '../../../../vscodeTypes';
import { createExtensionUnitTestingServices } from '../../../test/node/services';
import { GetErrorsTool } from '../getErrorsTool';
import { toolResultToString } from './toolTestUtils';

// Test the GetErrorsTool functionality
suite('GetErrorsTool - Tool Invocation', () => {
	let accessor: ITestingServicesAccessor;
	let collection: TestingServiceCollection;
	let diagnosticsService: TestLanguageDiagnosticsService;
	let tool: GetErrorsTool;

	const workspaceFolder = URI.file('/test/workspace');
	const tsFile1 = URI.file('/test/workspace/src/file1.ts');
	const tsFile2 = URI.file('/test/workspace/src/file2.ts');
	const jsFile = URI.file('/test/workspace/lib/file.js');
	const noErrorFile = URI.file('/test/workspace/src/noErrorFile.ts');
	const eslintErrorFile = URI.file('/test/workspace/eslint/eslint_unexpected_constant_condition_1.ts');

	beforeEach(() => {
		collection = createExtensionUnitTestingServices();

		// Set up test documents
		const tsDoc1 = createTextDocumentData(tsFile1, 'function test() {\n  const x = 1;\n  return x;\n}', 'ts').document;
		const tsDoc2 = createTextDocumentData(tsFile2, 'interface User {\n  name: string;\n  age: number;\n}', 'ts').document;
		const jsDoc = createTextDocumentData(jsFile, 'function legacy() {\n  var y = 2;\n  return y;\n}', 'js').document;
		const noErrorDoc = createTextDocumentData(noErrorFile, '', 'ts').document;
		const eslintErrorDoc = createTextDocumentData(eslintErrorFile, 'if (true) {\n  console.log("This is a constant condition");\n}', 'ts').document;

		collection.define(IWorkspaceService, new SyncDescriptor(TestWorkspaceService, [[workspaceFolder], [tsDoc1, tsDoc2, jsDoc, noErrorDoc, eslintErrorDoc]]));

		// Set up diagnostics service
		diagnosticsService = new TestLanguageDiagnosticsService();
		collection.define(ILanguageDiagnosticsService, diagnosticsService);

		accessor = collection.createTestingAccessor();

		// Create the tool instance
		tool = accessor.get(IInstantiationService).createInstance(GetErrorsTool);

		// Add test diagnostics
		diagnosticsService.setDiagnostics(tsFile1, [
			{
				message: 'Variable is declared but never used',
				range: new Range(1, 8, 1, 9),
				severity: DiagnosticSeverity.Warning
			},
			{
				message: 'Missing return type annotation',
				range: new Range(0, 9, 0, 13),
				severity: DiagnosticSeverity.Error
			}
		]);

		diagnosticsService.setDiagnostics(tsFile2, [
			{
				message: 'Interface should be exported',
				range: new Range(0, 0, 0, 9),
				severity: DiagnosticSeverity.Information // Should be filtered out
			},
			{
				message: 'Property age should be optional',
				range: new Range(2, 2, 2, 5),
				severity: DiagnosticSeverity.Warning
			}
		]);

		diagnosticsService.setDiagnostics(jsFile, [
			{
				message: 'Use const instead of var',
				range: new Range(1, 2, 1, 5),
				severity: DiagnosticSeverity.Warning
			}
		]);

		diagnosticsService.setDiagnostics(eslintErrorFile, [
			{
				message: 'Unexpected constant condition.',
				range: new Range(1, 4, 1, 4),
				severity: DiagnosticSeverity.Error
			}
		]);
	});

	afterEach(() => {
		accessor.dispose();
	});

	test('getDiagnostics - returns empty when no paths provided', () => {
		// Test getting all diagnostics
		const allDiagnostics = tool.getDiagnostics([]);
		expect(allDiagnostics).toEqual([]);
	});

	test('getDiagnostics - filters by file path', () => {
		// Test with specific file path
		const results = tool.getDiagnostics([{ uri: tsFile1, range: undefined }]);

		expect(results).toEqual([
			{ uri: tsFile1, diagnostics: diagnosticsService.getDiagnostics(tsFile1).filter(d => d.severity <= DiagnosticSeverity.Warning) } // Should only include Warning and Error
		]);
	});

	test('getDiagnostics - filters by folder path', () => {
		// Test with folder path
		const srcFolder = URI.file('/test/workspace/src');
		const results = tool.getDiagnostics([{ uri: srcFolder, range: undefined }]);

		// Should find diagnostics for files in the src folder
		expect(results).toEqual([
			{ uri: tsFile1, diagnostics: diagnosticsService.getDiagnostics(tsFile1).filter(d => d.severity <= DiagnosticSeverity.Warning) },
			{ uri: tsFile2, diagnostics: diagnosticsService.getDiagnostics(tsFile2).filter(d => d.severity <= DiagnosticSeverity.Warning) }
		]);
	});

	test('getDiagnostics - filters by range', () => {
		// Test with specific range that only covers line 1
		const range = new Range(1, 0, 1, 10);
		const results = tool.getDiagnostics([{ uri: tsFile1, range }]);

		expect(results).toEqual([
			{ uri: tsFile1, diagnostics: diagnosticsService.getDiagnostics(tsFile1).filter(d => d.severity <= DiagnosticSeverity.Warning && d.range.intersection(range)) }
		]);
	});

	test('getDiagnostics - file with no diagnostics returns empty diagnostics array', () => {
		const noErrorFile = URI.file('/test/workspace/src/noErrorFile.ts');
		const results = tool.getDiagnostics([{ uri: noErrorFile, range: undefined }]);

		expect(results).toEqual([
			{ uri: noErrorFile, diagnostics: [] }
		]);
	});

	// Tool invocation tests
	test('Tool invocation - with no filePaths aggregates all diagnostics and formats workspace message', async () => {
		const result = await tool.invoke({ input: {}, toolInvocationToken: null! }, CancellationToken.None);
		const msg = await toolResultToString(accessor, result);
		expect(msg).toMatchSnapshot();
	});

	test('Tool invocation - with single filePath limits diagnostics and message to that file', async () => {
		const pathRep = accessor.get(IPromptPathRepresentationService);
		const filePath = pathRep.getFilePath(tsFile1);
		const result = await tool.invoke({ input: { filePaths: [filePath] }, toolInvocationToken: null! }, CancellationToken.None);
		const msg = await toolResultToString(accessor, result);
		expect(msg).toMatchSnapshot();
	});

	test('Tool invocation - with folder path includes diagnostics from contained files', async () => {
		const pathRep = accessor.get(IPromptPathRepresentationService);
		const srcFolderUri = URI.file('/test/workspace/src');
		const srcFolderPath = pathRep.getFilePath(srcFolderUri);
		const result = await tool.invoke({ input: { filePaths: [srcFolderPath] }, toolInvocationToken: null! }, CancellationToken.None);
		const msg = await toolResultToString(accessor, result);
		expect(msg).toMatchSnapshot();
	});

	test('Tool invocation - with filePath and range filters diagnostics to that range', async () => {
		const pathRep = accessor.get(IPromptPathRepresentationService);
		const filePath = pathRep.getFilePath(tsFile1);
		// Range only covering the second line (line index 1) -> should include the warning at line 1 but not the error at line 0 if it doesn't intersect
		const range = new Range(1, 0, 1, 50);
		const result = await tool.invoke({
			input: {
				filePaths: [filePath],
				ranges: [[range.start.line, range.start.character, range.end.line, range.end.character]]
			},
			toolInvocationToken: null!
		}, CancellationToken.None);

		const msg = await toolResultToString(accessor, result);
		expect(msg).toMatchSnapshot();
	});

	test('Tool invocation - filePath with no diagnostics still has a <errors> entry', async () => {
		const pathRep = accessor.get(IPromptPathRepresentationService);
		const filePath = pathRep.getFilePath(noErrorFile);
		const result = await tool.invoke({ input: { filePaths: [filePath] }, toolInvocationToken: null! }, CancellationToken.None);
		const msg = await toolResultToString(accessor, result);
		expect(msg).toMatchSnapshot();
	});

	test('Tool invocation - filePath with range has a <compileError> entry', async () => {
		const pathRep = accessor.get(IPromptPathRepresentationService);
		const filePath = pathRep.getFilePath(eslintErrorFile);
		const result = await tool.invoke({ input: { filePaths: [filePath], ranges: [[1, 4, 1, 4]] }, toolInvocationToken: null! }, CancellationToken.None);
		const msg = await toolResultToString(accessor, result);
		expect(msg).toMatchSnapshot();
	});
});