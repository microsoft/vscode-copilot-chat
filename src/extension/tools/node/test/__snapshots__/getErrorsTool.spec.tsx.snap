// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GetErrorsTool - Tool Invocation > Tool invocation - filePath with no diagnostics still has a <errors> entry 1`] = `
"<errors path="/test/workspace/src/noErrorFile.ts">
No errors found
</errors>
"
`;

exports[`GetErrorsTool - Tool Invocation > Tool invocation - filePath with range has a <compileError> entry 1`] = `
"<errors path="/test/workspace/eslint/eslint_unexpected_constant_condition_1.ts">
This code at line 2
\`\`\`
  console.log("This is a constant condition");
\`\`\`
has the problem reported:
<compileError>
Unexpected constant condition.
</compileError>

</errors>
"
`;

exports[`GetErrorsTool - Tool Invocation > Tool invocation - with filePath and range filters diagnostics to that range 1`] = `
"<errors path="/test/workspace/src/file1.ts">
This code at line 2
\`\`\`
  const x = 1;
\`\`\`
has the problem reported:
<compileError>
Variable is declared but never used
</compileError>

</errors>
"
`;

exports[`GetErrorsTool - Tool Invocation > Tool invocation - with folder path includes diagnostics from contained files 1`] = `
"<errors path="/test/workspace/src/file1.ts">
This code at line 2
\`\`\`
  const x = 1;
\`\`\`
has the problem reported:
<compileError>
Variable is declared but never used
</compileError>
This code at line 1
\`\`\`
function test() {
\`\`\`
has the problem reported:
<compileError>
Missing return type annotation
</compileError>

</errors>
<errors path="/test/workspace/src/file2.ts">
This code at line 3
\`\`\`
  age: number;
\`\`\`
has the problem reported:
<compileError>
Property age should be optional
</compileError>

</errors>
"
`;

exports[`GetErrorsTool - Tool Invocation > Tool invocation - with no filePaths aggregates all diagnostics and formats workspace message 1`] = `
"<errors path="/test/workspace/src/file1.ts">
This code at line 2
\`\`\`
  const x = 1;
\`\`\`
has the problem reported:
<compileError>
Variable is declared but never used
</compileError>
This code at line 1
\`\`\`
function test() {
\`\`\`
has the problem reported:
<compileError>
Missing return type annotation
</compileError>

</errors>
<errors path="/test/workspace/src/file2.ts">
This code at line 3
\`\`\`
  age: number;
\`\`\`
has the problem reported:
<compileError>
Property age should be optional
</compileError>

</errors>
<errors path="/test/workspace/lib/file.js">
This code at line 2
\`\`\`
  var y = 2;
\`\`\`
has the problem reported:
<compileError>
Use const instead of var
</compileError>

</errors>
<errors path="/test/workspace/eslint/eslint_unexpected_constant_condition_1.ts">
This code at line 2
\`\`\`
  console.log("This is a constant condition");
\`\`\`
has the problem reported:
<compileError>
Unexpected constant condition.
</compileError>

</errors>
"
`;

exports[`GetErrorsTool - Tool Invocation > Tool invocation - with single filePath limits diagnostics and message to that file 1`] = `
"<errors path="/test/workspace/src/file1.ts">
This code at line 2
\`\`\`
  const x = 1;
\`\`\`
has the problem reported:
<compileError>
Variable is declared but never used
</compileError>
This code at line 1
\`\`\`
function test() {
\`\`\`
has the problem reported:
<compileError>
Missing return type annotation
</compileError>

</errors>
"
`;
